src/chunked/encoder.rs

```diff
@@ -0,0 +1,185 @@
+use std::pin::Pin;
+
+use async_std::io;
+use async_std::io::prelude::*;
+use async_std::task::{Context, Poll};
+use http_types::Response;
+
+const CR: u8 = b'\r';
+const LF: u8 = b'\n';
+
+/// The encoder state.
+#[derive(Debug)]
+enum State {
+    /// Streaming out chunks.
+    Streaming,
+    /// Receiving trailers from a channel.
+    ReceiveTrailers,
+    /// Streaming out trailers.
+    EncodeTrailers,
+    /// Writing the final CRLF
+    EndOfStream,
+    /// The stream has finished.
+    Done,
+}
+
+/// An encoder for chunked encoding.
+#[derive(Debug)]
+pub(crate) struct ChunkedEncoder {
+    /// How many bytes we've written to the buffer so far.
+    bytes_written: usize,
+    /// The internal encoder state.
+    state: State,
+}
+
+impl ChunkedEncoder {
+    /// Create a new instance.
+    pub(crate) fn new() -> Self {
+        Self {
+            state: State::Streaming,
+            bytes_written: 0,
+        }
+    }
+    /// Encode an AsyncBufRead using "chunked" framing. This is used for streams
+    /// whose length is not known up front.
+    ///
+    /// # Format
+    ///
+    /// Each "chunk" uses the following encoding:
+    ///
+    /// ```txt
+    /// 1. {byte length of `data` as hex}\r\n
+    /// 2. {data}\r\n
+    /// ```
+    ///
+    /// A chunk stream is finalized by appending the following:
+    ///
+    /// ```txt
+    /// 1. 0\r\n
+    /// 2. {trailing header}\r\n (can be repeated)
+    /// 3. \r\n
+    /// ```
+    pub(crate) fn encode(
+        &mut self,
+        res: &mut Response,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        self.bytes_written = 0;
+        match self.state {
+            State::Streaming => self.encode_stream(res, cx, buf),
+            State::ReceiveTrailers => self.encode_trailers(res, cx, buf),
+            State::EncodeTrailers => self.encode_trailers(res, cx, buf),
+            State::EndOfStream => self.encode_eos(cx, buf),
+            State::Done => Poll::Ready(Ok(0)),
+        }
+    }
+
+    /// Stream out data using chunked encoding.
+    fn encode_stream(
+        &mut self,
+        mut res: &mut Response,
+        cx: &mut Context<'_>,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        // Get bytes from the underlying stream. If the stream is not ready yet,
+        // return the header bytes if we have any.
+        let src = match Pin::new(&mut res).poll_fill_buf(cx) {
+            Poll::Ready(Ok(n)) => n,
+            Poll::Ready(Err(e)) => return Poll::Ready(Err(e)),
+            Poll::Pending => match self.bytes_written {
+                0 => return Poll::Pending,
+                n => return Poll::Ready(Ok(n)),
+            },
+        };
+
+        // If the stream doesn't have any more bytes left to read we're done.
+        if src.len() == 0 {
+            // Write out the final empty chunk
+            let idx = self.bytes_written;
+            buf[idx] = b'0';
```

Comment:
Is it possible for `buf.len() < 3`?

Reply:
Yeah was wondering about that too. It shouldn't be. But we can probably add a `debug_assert` so it'll panic with a better message.

Reply:
Actually: I think we should wake the waker and return bytes written so far. Good one!